esa/esa.go

```diff
@@ -0,0 +1,312 @@
+// Package esa provides a client library for esa API v1
+package esa
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	baseURL    = "https://api.esa.io/"
+	apiVersion = "v1"
+
+	// RateLimit Headers
+	// ref. https://docs.esa.io/posts/102#2-3-0
+	headerRateLimit     = "X-RateLimit-Limit"
+	headerRateRemaining = "X-RateLimit-Remaining"
+	headerRateReset     = "X-RateLimit-Reset"
+)
+
+// A Client manages communication with the esa API v1.
+type Client struct {
+	clientMu sync.Mutex // clientMu protects the client during calls that modify the CheckRedirect func.
+
+	client  *http.Client
+	BaseURL *url.URL
+
+	rateMu    sync.Mutex
+	rateLimit Rate // Rate limit for the client as determined by the most recent API calls.
+
+	common service // Reuse a single struct instead of allocating one for each service on the heap.
```

Comment:
**[unused]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
field common is unused (U1000)